src/main/java/dev/lu15/voicechat/VoiceChatImpl.java

```diff
@@ -116,31 +131,124 @@ private void handle(@NotNull Player player, @NotNull HandshakePacket packet) {
             player.sendPacket(this.packetHandler.write(new HandshakeAcknowledgePacket(
                     event.getSecret(),
                     this.port,
-                    player.getUuid(), // why is this sent? the client already knows the player's uuid
-                    Codec.VOIP, // todo: configurable
-                    1024, // todo: configurable
-                    48, // todo: configurable
-                    1000, // todo: configurable
-                    false, // todo: configurable
+                    player.getUuid(),
+                    codec,
+                    mtu,
+                    distance,
+                    keepalive,
+                    groups,
                     this.publicAddress,
-                    false // todo: configurable
+                    recording // todo: configurable (recording)
             )));
+            groupManager.groups.forEach((id, group) -> {
+                player.sendPacket(this.packetHandler.write(new GroupCreatedPacket(group)));
+            });
         });
     }
 
     private void handle(@NotNull Player player, @NotNull UpdateStatePacket packet) {
         // todo: set state when players disconnect from voice chat server - NOT when they disconnect from the minecraft server
+        VoiceState oldstate = player.getTag(Tags.PLAYER_STATE);
+        UUID group = null;
+        if(oldstate!=null&&oldstate.group()!=null) {
```

Comment:
This condition is pretty unreadable, proper spacing should be used in statements. This comment probably goes for the rest of the added statements across the PR.